### PR TITLE
Removed default page transitions for desktop and web platforms.

### LINKED ISSUE #82053: Change default material page transition for Web and Desktop
## Use case

Currently the default material page transition is a "slide transition" where the new screen comes from the right and pushes on top of the screen you are currently on.

On mobile this a common screen transition paradigm that fits the expectation of the user and the screen aspect ratio and format of a mobile device.

However, websites do not typically have this type of transition effect. Normally when you click a link or transition to another feature or feature areas there is no transition at all. The page just changes and the content is replaced.

So, when building a Flutter web (or desktop) app the default transition seems unexpected to the user and negatively affects user experience.

## Proposal

On web and desktop targets, change the default material page transition to be no animation just a page replacement.

### PR DESCRIPTION
**Pull request by darrenaustin**
The current default page transitions defined for desktop and web platforms use existing slide-up or fade transitions used on Android and iOS. As page transitions like this are not the norm for desktops or web, this PR changes the defaults to only have transitions for the mobile platforms and shuts them off for everything else including web running on anything.

This doesn't appear to be a breaking change as I tested it against the community tests and our internal tests. However, if you need to restore the previous defaults, you can add this to your app's theme:

```dart
MaterialApp(
  theme: ThemeData(
      pageTransitionsTheme: const PageTransitionsTheme(
        builders: <TargetPlatform, PageTransitionsBuilder>{
          TargetPlatform.android: FadeUpwardsPageTransitionsBuilder(),
          TargetPlatform.iOS: CupertinoPageTransitionsBuilder(),
          TargetPlatform.linux: FadeUpwardsPageTransitionsBuilder(),
          TargetPlatform.macOS: CupertinoPageTransitionsBuilder(),
          TargetPlatform.windows: FadeUpwardsPageTransitionsBuilder(),
        },
      )
  )
)
```

Fixes: #82053

I updated several tests to ensure the right transitions are getting picked up, as well as removed `macOS` from several tests and turned some off when running on the web, as transitions were no longer the default in those configurations.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt.
- [x] All existing and new tests are passing.

### CODE REVIEW

File: packages/flutter/lib/src/material/page_transitions_theme.dart
Comment (goderbauer) [inline packages/flutter/lib/src/material/page_transitions_theme.dart:596]: nit: link from the builder property to this one and vice versa?

**Review by goderbauer — COMMENTED** [review]
Wondering: Does the material spec define any kind of transition for Desktop (or the web)? In other words: Is not having a transition in line with the spec?

/cc @yjbanov Is it the right thing to not have a transition on the web?

File: packages/flutter/lib/src/material/page_transitions_theme.dart
Comment (chunhtai) [inline packages/flutter/lib/src/material/page_transitions_theme.dart:576]: This API feels weird to me, when you build for the web you usually don't care about the native platforms. You probably care about browsers, but that is not the case for the flutter.

I wonder whether we can change this API so that there is just one builder for the web, and then makes the API easier to expand to differentiate each platform if customers really want to do that.

**Comment by darrenaustin** [comment]
> Wondering: Does the material spec define any kind of transition for Desktop (or the web)? In other words: Is not having a transition in line with the spec?

I looked, but didn't see any guidance for this. As this is an uncommon in desktop applications it seems reasonable to turn them off. As I showed above, if and app does want them it is pretty straightforward to add them to the theme data for the app.

File: packages/flutter/lib/src/material/page_transitions_theme.dart
Comment (darrenaustin) [reply packages/flutter/lib/src/material/page_transitions_theme.dart:576]: > I wonder whether we can change this API so that there is just one builder for the web, and then makes the API easier to expand to differentiate each platform if customers really want to do that.

I did consider that, but we wouldn't be able to add it to the existing `builders` field as there isn't a `TargetPlatform.web`. So if we have to add another field for the web, why not provide a way to specialize it for the underlying host platform? That way if you are writing a PWA and want it to be as native looking as possible you could provide the native transitions for each platform with the `webBuilders` field.

Comment (chunhtai) [reply packages/flutter/lib/src/material/page_transitions_theme.dart:576]: Not sure whether this is a good idea, just want to throw options around. Can we do this as a bug fix without adding new API? for example we conditionally sets the _builders to _defaultWebBuilders if kIsWeb is true. This way, the behavior is the same, and user can still customize using the old api by check the kIsWeb themselves. We don't risk exposing a new API that we aren't sure whether it can be useful.

Comment (goderbauer) [reply packages/flutter/lib/src/material/page_transitions_theme.dart:576]: That sounds good to me. However, would probably be good to get some inout from the web team around whether defaulting to no transition is the right thing (I think it is).

**Comment by darrenaustin** [comment]
I just updated the PR based on the review comments. I switched back to just a single `builders` parameter to the theme to keep it simple. Please let me know what you think. Thx.

**Review by chunhtai — APPROVED** [review]
LGTM if also LGT web team

**Review by goderbauer — APPROVED** [review]
LGTM

**Comment by yjbanov** [comment]
Our designers tell me that full page transitions without animations make sense, so LGTM!

**Review by yjbanov — APPROVED** [review]
lgtm

File: packages/flutter/lib/src/material/page_transitions_theme.dart
Comment (yjbanov) [inline packages/flutter/lib/src/material/page_transitions_theme.dart:572]: `kIsWeb` is true both in desktop and mobile browsers. Do we want to shut off transitions in mobile browsers too?

Comment (darrenaustin) [reply packages/flutter/lib/src/material/page_transitions_theme.dart:572]: I believe we do. That is the feedback I got when I asked this question on this [issue](https://github.com/flutter/flutter/issues/82053#issuecomment-841457312). For most web apps that aren't PWAs I would think they would want consistent behavior. For apps that do want transitions, they can provide them either manually or by using the theme settings detailed above.

**Comment by johnpryan** [comment]
I disagree that this should be disabled for the entire web platform - many web apps would like to keep the page transitions when the user accesses the web app from a mobile browser.

Could we base this decision on the platform, not the kIsWeb flag instead?

**Comment by darrenaustin** [comment]
> I disagree that this should be disabled for the entire web platform - many web apps would like to keep the page transitions when the user accesses the web app from a mobile browser.

You can get this behaviour by adding the following to their app's theme:

```dart
MaterialApp(
  theme: ThemeData(
      pageTransitionsTheme: const PageTransitionsTheme(
        builders: <TargetPlatform, PageTransitionsBuilder>{
          TargetPlatform.android: FadeUpwardsPageTransitionsBuilder(),
          TargetPlatform.iOS: CupertinoPageTransitionsBuilder(),
        },
      )
  )
)
```

Which will override the defaults for both web and mobile to use the mobile style transitions.

This PR is just changing the defaults, and the feedback we got was that in general apps for the web did not want transitions by default.

**Comment by goderbauer** [comment]
> You can get this behaviour by adding the following to their app's theme:

We should add that to the API docs somewhere. Maybe even on MaterialApp?

**Pull request merged**
